src/System.Linq.Expressions/tests/ILReader/SigParser.cs

```diff
@@ -0,0 +1,941 @@
+                    NotifyTypeClass();
+
+                    sig_index encodedC;
+
+                    if (!ParseTypeDefOrRefEncoded(out encodedC, out indexType, out index))
+                        return false;
+
+                    NotifyTypeDefOrRef(encodedC, indexType, index);
+                    break;
+
+                case ELEMENT_TYPE_VALUETYPE:
+                    //VALUETYPE TypeDefOrRefEncoded
+                    NotifyTypeValueType();
+
+                    sig_index encodedV;
+
+                    if (!ParseTypeDefOrRefEncoded(out encodedV, out indexType, out index))
+                        return false;
+
+                    NotifyTypeDefOrRef(encodedV, indexType, index);
+                    break;
+
+                case ELEMENT_TYPE_FNPTR:
+                    // FNPTR MethodDefSig
+                    // FNPTR MethodRefSig
+                    NotifyTypeFunctionPointer();
+
+                    if (!ParseByte(out elem_type))
+                        return false;
+
+                    if (!ParseMethod(elem_type))
+                        return false;
+
+                    break;
+
+                case ELEMENT_TYPE_ARRAY:
+                    // ARRAY Type ArrayShape
+                    NotifyTypeArray();
+
+                    if (!ParseType())
+                        return false;
+
+                    if (!ParseArrayShape())
+                        return false;
+                    break;
+
+                case ELEMENT_TYPE_SZARRAY:
+                    // SZARRAY CustomMod* Type
+
+                    NotifyTypeSzArray();
+
+                    if (!ParseOptionalCustomMods())
+                        return false;
+
+                    if (!ParseType())
+                        return false;
+
+                    break;
+
+                case ELEMENT_TYPE_GENERICINST:
+                    // GENERICINST (CLASS | VALUETYPE) TypeDefOrRefEncoded GenArgCount Type *
+
+                    if (!ParseType())
+                        return false;
+
+                    if (!ParseNumber(out number))
+                        return false;
+
+                    for (sig_mem_number i = 0; i < number; i++)
+                    {
+                        if (!ParseType())
+                            return false;
+                    }
+
+                    NotifyTypeGenericInst(number);
+
+                    break;
+
+                case ELEMENT_TYPE_VAR:
+                    // VAR Number
+                    if (!ParseNumber(out number))
+                        return false;
+                    NotifyTypeGenericTypeVariable(number);
+                    break;
+
+                case ELEMENT_TYPE_MVAR:
+                    // MVAR Number
+                    if (!ParseNumber(out number))
+                        return false;
+                    NotifyTypeGenericMemberVariable(number);
+                    break;
+
+                case ELEMENT_TYPE_INTERNAL:
```

Comment:
This is the case that prevents me from using https://github.com/dotnet/corefx/pull/4809.